client/src/test-runner/result.ts

```diff
@@ -0,0 +1,278 @@
+        data: dataMap,
+      };
+    }
+  } else if (failure.reason.data) {
+    return {
+      tag: "message",
+      message: String(failure.reason.data),
+    };
+  } else if (failure.message) {
+    return {
+      tag: "message",
+      message: String(failure.message),
+    };
+  }
+  throw new Error(`unknown failure ${JSON.stringify(failure)}`);
+}
+
+export function parseErrorOutput(line: string): ErrorOutput {
+  const errors: json.ParseError[] = [];
+  const output: CompileErrors = json.parse(line, errors);
+  const nojson = errors.find(
+    (e) => e.error === json.ParseErrorCode.InvalidSymbol,
+  );
+  if (errors.length > 0 && nojson) {
+    return { type: "message", line };
+  }
+  return output;
+}
+
+export type Output = Message | Result;
+
+export type ErrorOutput = Message | CompileErrors;
+
+export type Message = {
+  type: "message";
+  line: string;
+};
+
+export type Result = {
+  type: "result";
+  event: Event;
+};
+
+export type Event =
+  | { tag: "runStart"; testCount: number }
+  | EventTestCompleted
+  | { tag: "runComplete"; passed: number; failed: number; duration: number };
+
+export type EventTestCompleted = {
+  tag: "testCompleted";
+  labels: string[];
+  messages: string[];
+  duration: number;
+  status: TestStatus;
+};
+
+export type TestStatus =
+  | { tag: "pass" }
+  | { tag: "todo"; comment: string }
+  | { tag: "fail"; failures: Failure[] };
+
+export type Failure =
+  | { tag: "message"; message: string }
+  | {
+      tag: "comparison";
+      comparison: string;
+      actual: string;
+      expected: string;
+    }
+  | { tag: "data"; data: { [key: string]: string } };
+
+export type CompileErrors = {
+  type: "compile-errors";
+  errors: Error[];
+};
+
+export type Error = {
+  path: string;
+  name: string;
+  problems: Problem[];
+};
+
+export type Problem = {
+  title: string;
+  region: Region;
+  message: MessagePart[];
+};
+
+export type Region = {
+  start: Position;
+  end: Position;
+};
+
+export type Position = {
+  line: number;
+  column: number;
+};
+
+export type MessagePart = string | StyledString;
+
+export type StyledString = {
+  bold?: boolean;
+  underline?: boolean;
+  color?: string;
+  string: string;
+};
+
+export function buildMessage(event: EventTestCompleted): string | undefined {
+  if (event.status.tag === "fail") {
+    const lines = event.status.failures.flatMap((failure) => {
+      switch (failure.tag) {
+        case "comparison":
+          return [failure.actual, "| " + failure.comparison, failure.expected];
+        case "data":
+          return Object.keys(failure.data).map(
+            (key) => `${key}: ${failure.data[key]}`,
+          );
+        case "message":
+          return [failure.message];
+      }
+    });
+    return event.messages.concat(lines).join("\n");
+  }
+  return event.messages.join("\n");
+}
+
+function evalStringLiteral(value: string): string {
+  if (value && value.startsWith('"')) {
+    return String(eval(value));
```

Comment:
What are we doing this for?

In my test case it tries to eval "Just Login", which javascript probably doesn't evaluate correctly anyway?